keps/672-serial-job-execution/README.md

```diff
@@ -0,0 +1,500 @@
+)
+```
+
+### Implementation
+
+The JobSet operator will control the creation of ReplicatedJobs based on their DependsOn
+configuration.
+
+In the DependsOn API, user can only reference the **single** ReplicatedJob that previously
+defined in the `.spec.replicatedJobs` list.
+
+If ReplicatedJob has the DependsOn configuration, controller will check the counter of
+Ready or Complete Jobs in the referenced ReplicatedJob. When the counter of Jobs is equal to
+referenced ReplicatedJob's replica count, the controller will create the ReplicatedJob.
+
+If JobSet is suspended the all ReplicatedJobs will be suspended and the Job sequence starts again.
+
+When the JobSet is restarted after failure, the Job sequence starts again. User controls how many
+times Job can be restarted via backOffLimit parameter.
+
+### Quota Management
+
+In the initial implementation of the DependsOn API the resource quota will be calculated as
+sum of all ReplicatedJobs resources. Which means JobSet will be admitted by
+[Kueue](https://github.com/kubernetes-sigs/kueue) only when all resources are available for
+every ReplicatedJob within JobSet.
+
+That allows us to leverage the existing integration between Kueue and JobSet while using the
+DependsOn API.
+
+In the future versions we will discuss how Kueue can enqueue group of ReplicatedJobs separately
+and admit them. For example, when compute resources are available for the first ReplicatedJob
+the JobSet can be dispatched by Kueue.
+
+### Deprecation of StartupPolicy
+
+Since the new DependsOn API covers use-cases for the `startupPolicy: InOrder`, we will deprecate the
+StartupPolicy API in the future version of JobSet. Before deprecation, we will make
+StartupPolicy and DependsOn API mutually exclusive. Which means, user can use only
+StartupPolicy or DependsOn API.
+
+User should configure JobSet as follows to use the existing functionality of StartupPolicy:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: mpi-job
+spec:
+  replicatedJobs:
+    - name: launcher
+    - name: trainer-node
+      dependsOn:
+        - name: launcher
+          status: Ready
+```
+
+### Defaulting/Validation
+
+- DependsOn API is immutable.
+- Length of DependsOn list is equal to 1.
+- Ensure that `replicatedJobs[n].dependsOn[0].name` is equal to the the previously defined
+  ReplicatedJob's name (e.g. `n-1` from the list).
+- If DependsOn is set, `name` and `status` must be configured.
+- The first ReplicatedJob (e.g `n=0` in the list) can't contain DependsOn API.
+- Make DependsOn and StartupPolicy mutually exclusive.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+#### Unit Tests
+
+- `controllers`: `01/19/2024` - `30.2%`
+
+We will add this functionality to the startup_policy API for the functionality.
```

Comment:
If we are going to deprecate startup policy I would probably put the execution policy elsewhere.

Reply:
What do you mean by that ?
Similar to what @ahg-g proposed above, but with the executionPolicy field name: https://github.com/kubernetes-sigs/jobset/pull/680#discussion_r1875924692 ?

Reply:
We will add this functionality to the startup_policy API for the functionality.

You are going to put the code in startup_policy.go but we are going to deprecate startup_policy. That seems odd to me.

Reply:
Oh, I think, we should put the code under `depends_on.go` or `execution_policy.go` file, but I am open to the various options.

Reply:
Yea I think depends_on.go is fine but that is mostly an implementation detail.